### PR TITLE
[#54] Improve documentation

### LINKED ISSUE #54: Improve docs
- [ ] Walk through functions add usage examples and description if needed
- [ ] Fix existing docs where necessary (if any mistakes)
- [ ] Fix module headers

### PR DESCRIPTION
**Pull request by ChShersh**
Resolves #54 

See `CHANGELOG.md` for list of changes. Some of them are breaking. But that's okay, since we're going to release new major version anyways.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -208,7 +208,11 @@ We also have `data Undefined = Undefined` (which, too, comes with warnings).
 
 ### Exceptions
 
-TODO: write about reexports, `Bug` and `Exc` pattern.
+`relude` reexports `Exception` type from `base` package and introduces `bug`
```
Comment (vrom911) [inline README.md]: I would add `the` before `base` and `bug`

File: README.md [hunk 2]
```diff
@@ -208,7 +208,11 @@ We also have `data Undefined = Undefined` (which, too, comes with warnings).
...
+function as an alternative to `error`. There's also very convenient `Exc`
```
Comment (vrom911) [inline README.md]: *a* very

File: src/Relude/Base.hs [hunk 3]
```diff
@@ -76,12 +74,13 @@ import Data.Proxy (Proxy (..))
...
 
-import GHC.Base (String, asTypeOf, maxInt, minInt, ord, seq)
+import GHC.Base (String, asTypeOf, maxInt, minInt, ord, seq, ($!))
 import GHC.Enum (Bounded (..), Enum (..), boundedEnumFrom, boundedEnumFromThen)
 import GHC.Float (Double (..), Float (..), Floating (acos, acosh, asin, asinh, atan, atanh, cos, cosh, exp, logBase, pi, sin, sinh, sqrt, tan, tanh, (**)))
 import GHC.Generics (Generic)
 import GHC.Num (Integer, Num (..), subtract)
-import GHC.Real hiding (showSigned, (%))
+import GHC.Real (Fractional (..), Integral (..), Ratio, Rational, Real (..), RealFrac (..), even,
+                 fromIntegral, gcd, lcm, odd, realToFrac, (^), (^^))
```
Comment (vrom911) [inline src/Relude/Base.hs]: What about `denominator` and `numerator`?

File: src/Relude/Extra/Enum.hs [hunk 4]
```diff
@@ -26,7 +35,12 @@ import qualified Data.Map.Strict as M
 universe :: (Bounded a, Enum a) => [a]
 universe = [minBound .. maxBound]
 
-{- | Creates a function that is the inverse of a given function @f@.
+{- | @inverseMap f@ creates a function that is the inverse of a given function
+@f@. It does so by constructing 'M.Map' for every value @f a@. Implementation
```
Comment (vrom911) [inline src/Relude/Extra/Enum.hs]: _The_ implementation

File: src/Relude/Extra/Map.hs [hunk 5]
```diff
@@ -1,7 +1,14 @@
 {-# LANGUAGE TypeFamilies #-}
 
-{- | Contains implememtation of polymorhic type classes for things like 'Set'
-and 'Map'.
+{- |
+Copyright:  (c) 2018 Kowainik
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
+
+Polymorphic grouping functions.
```
Comment (vrom911) [inline src/Relude/Extra/Map.hs]: This is redundant.

File: src/Relude/Extra/Map.hs [hunk 6]
```diff
@@ -1,7 +1,14 @@
...
+
+Contains implememtation of polymorhic type classes for things like 'Set' and
```
Comment (vrom911) [inline src/Relude/Extra/Map.hs]: typos: _the implementation of polymorphic_

File: src/Relude/Extra/Map.hs [hunk 7]
```diff
@@ -1,7 +1,14 @@
...
+'Map'.
 -}
 
 module Relude.Extra.Map
```
Comment (vrom911) [inline src/Relude/Extra/Map.hs]: at line 48 here also typo: _polymorphic_

File: src/Relude/Foldable/Fold.hs [hunk 8]
```diff
@@ -35,41 +46,47 @@ import Relude.Container.Reexport (HashSet, Set)
...
 -}
 flipfoldl' :: Foldable f => (a -> b -> b) -> b -> f a -> b
 flipfoldl' f = foldl' (flip f)
 {-# INLINE flipfoldl' #-}
 
-foldMapA :: (Monoid b, Applicative m, Foldable f) => (a -> m b) -> f a -> m b
+{- | Polymorphic version of @concatMapA@ function.
+
+>>> foldMapA @[Int] (Just . replicate 3) [1..3]
+Just [1,1,1,2,2,2,3,3,3]
```
Comment (vrom911) [inline src/Relude/Foldable/Fold.hs]: 👍

File: src/Relude/Foldable/Fold.hs [hunk 9]
```diff
@@ -35,41 +46,47 @@ import Relude.Container.Reexport (HashSet, Set)
...
+foldMapA :: forall b m f a . (Monoid b, Applicative m, Foldable f) => (a -> m b) -> f a -> m b
 foldMapA f = foldr step (pure mempty)
   where
     step a mb = mappend <$> f a <*> mb
 {-# INLINE foldMapA #-}
 
-foldMapM :: (Monoid b, Monad m, Foldable f) => (a -> m b) -> f a -> m b
+{- | Polymorphic version of @concatMapM@ function.
+
+>>> foldMapA @[Int] (Just . replicate 3) [1..3]
```
Comment (vrom911) [inline src/Relude/Foldable/Fold.hs]: The example is for `foldMapA`

File: src/Relude/Lifted/Concurrent.hs [hunk 10]
```diff
@@ -1,13 +1,14 @@
...
 Copyright: (c) 2016 Stephen Diehl
            (c) 20016-2018 Serokell
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
--- | Concurrency useful and common functions.
+Concurrency useful and common functions.
```
Comment (vrom911) [inline src/Relude/Lifted/Concurrent.hs]: I would rephrase it

File: src/Relude/Functor/Reexport.hs [hunk 11]
```diff
@@ -1,13 +1,14 @@
...
 Copyright: (c) 2016 Stephen Diehl
            (c) 20016-2018 Serokell
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
--- | This module reexports functionality regarding 'Functor' type class.
+This module reexports functionality regarding 'Functor' type class.
```
Comment (vrom911) [inline src/Relude/Functor/Reexport.hs]: I noticed that short style description is looking better. Probably we should consider replacing with something like this:
_ Functionality for 'Functor' type class._
What do you like more?

File: src/Relude/Monad/Reexport.hs [hunk 12]
```diff
@@ -1,13 +1,14 @@
...
 Copyright: (c) 2016 Stephen Diehl
            (c) 20016-2018 Serokell
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
--- | This module reexports functions to work with monads.
+This module reexports functions to work with monads.
```
Comment (vrom911) [inline src/Relude/Monad/Reexport.hs]: Reexports to work with Monads?

File: src/Relude/Monoid.hs [hunk 13]
```diff
@@ -1,11 +1,12 @@
...
 Copyright: (c) 2016 Stephen Diehl
            (c) 20016-2018 Serokell
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
--- | This module reexports functions to work with monoids plus adds extra useful functions.
+This module reexports functions to work with monoids plus adds extra useful functions.
```
Comment (vrom911) [inline src/Relude/Monoid.hs]: Functions to work with Monoids.

File: src/Relude/String/Reexport.hs [hunk 14]
```diff
@@ -1,11 +1,12 @@
...
 Copyright: (c) 2016 Stephen Diehl
            (c) 20016-2018 Serokell
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
--- | This module reexports functions to work with 'Text' and 'ByteString' types.
+This module reexports functions to work with 'Text' and 'ByteString' types.
```
Comment (vrom911) [inline src/Relude/String/Reexport.hs]: Functions o work with 'Text' and 'ByteString' types.

File: src/Relude/Unsafe.hs [hunk 15]
```diff
@@ -1,13 +1,13 @@
...
            (c) 20016-2018 Serokell
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
-{- | Unsafe functions to work with lists and 'Maybe'.
+Unsafe functions to work with lists and 'Maybe'.
 Sometimes unavoidable but better don't use them. This module
```
Comment (vrom911) [inline src/Relude/Unsafe.hs]: probably it's better to say: "but it's better not to use them"?

File: src/Relude.hs [hunk 16]
```diff
@@ -1,49 +1,63 @@
...
+@[base-noprelude](http://hackage.haskell.org/package/base-noprelude)@ and add
+the following 'Prelude' module to your package to use @relude@ by default in
+every module istead of 'Prelude':
+
+@
+__module__ Prelude (__module__ "Relude") __where__
+__import__ "Relude"
+@
+
 This documentation section contains description of internal module structure to
```
Comment (vrom911) [inline src/Relude.hs]: the description of the internal

File: src/Relude.hs [hunk 17]
```diff
@@ -1,49 +1,63 @@
...
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
-{- | Main module that reexports all functionality allowed to use
-without importing any other modules. Just add next lines to your
-module to replace default 'Prelude' with better one.
+Main module that reexports all functionality allowed to use without importing
```
Comment (vrom911) [inline src/Relude.hs]: The main module that reexports all functionality. It's allowed to use it..

File: src/Relude.hs [hunk 18]
```diff
@@ -1,49 +1,63 @@
...
 @
 \{\-\# LANGUAGE NoImplicitPrelude \#\-\}
 
 __import__ "Relude"
 @
 
+Alternatively, you can replace @base@ package in your dependencies with
+@[base-noprelude](http://hackage.haskell.org/package/base-noprelude)@ and add
+the following 'Prelude' module to your package to use @relude@ by default in
+every module istead of 'Prelude':
```
Comment (vrom911) [inline src/Relude.hs]: _instead_

File: src/Relude.hs [hunk 19]
```diff
@@ -1,49 +1,63 @@
...
 This documentation section contains description of internal module structure to
 help navigate between modules, search for interesting functionalities and
-understand where you need to put your new changes.
+understand where you need to put your new changes (if you're a contributor).
 
 Functions and types are distributed across multiple modules and grouped by
-meaning or __theme__. Name of the module should give you hints regarding what
-this module contains. Some /themes/ contain a great amount of both reexported
+meaning or __category__. Name of the module should give you hints regarding what
+this module contains. Some /categories/ contain a great amount of both reexported
```
Comment (vrom911) [inline src/Relude.hs]: great -> significant

File: src/Relude.hs [hunk 20]
```diff
@@ -1,49 +1,63 @@
...
 functions and functions of our own. To make it easier to understand these huge
```
Comment (vrom911) [inline src/Relude.hs]: huge -> enormous

File: src/Relude.hs [hunk 21]
```diff
@@ -1,49 +1,63 @@
...
 chunks of functions, all reexported stuff is moved into separate module with
```
Comment (vrom911) [inline src/Relude.hs]: _the_ separate

File: src/Relude/Bool/Guard.hs [hunk 22]
```diff
@@ -1,11 +1,12 @@
...
 Copyright: (c) 2016 Stephen Diehl
            (c) 20016-2018 Serokell
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
--- | This module contains monadic predicates.
+This module contains monadic boolean combinators.
```
Comment (vrom911) [inline src/Relude/Bool/Guard.hs]: just "Monadic boolean combinators"

File: src/Relude/Debug.hs [hunk 23]
```diff
@@ -10,16 +10,17 @@
...
            (c) 2018 Kowainik
-License: MIT
--}
+License:    MIT
+Maintainer: Kowainik <xrom.xkov@gmail.com>
 
--- | Functions for debugging. If you left these functions in your code
--- then warning is generated to remind you about left usages. Also some
--- functions (and data types) are convenient for prototyping.
+Functions for debugging. If you left these functions in your code then warning
```
Comment (vrom911) [inline src/Relude/Debug.hs]: _a_ warning

File: src/Relude/Extra/CallStack.hs [hunk 24]
```diff
@@ -24,4 +27,21 @@ that the caller function has 'HasCallStack' constraint. Otherwise, it returns
 ownName :: HasCallStack => String
 ownName = case getCallStack callStack of
     _:caller:_ -> fst caller
-    _          -> "ownName"
+    _          -> "<unknown>"
+
+{- | This function returns the name of its caller of the caller function, but it
+requires that the caller function and caller of the caller function has
```
Comment (vrom911) [inline src/Relude/Extra/CallStack.hs]: has -> have

File: src/Relude/List/Reexport.hs [hunk 25]
```diff
@@ -7,7 +7,7 @@ Copyright: (c) 2016 Stephen Diehl
 License:    MIT
 Maintainer: Kowainik <xrom.xkov@gmail.com>
 
-This module reexports functinons to work with list and 'NonEmpty'.
+Reexports most of the "Data.List" and "Data.List.NonEmpty".
```
Comment (vrom911) [inline src/Relude/List/Reexport.hs]: Are these "" are intentional?

Comment (ChShersh) [reply src/Relude/List/Reexport.hs]: @vrom911 yes, they give links to modules. In Haddock we need to use single quotes to link to function or types and double quotes to link to modules.
